### ISSUE TITLE
JQuery wrong types

### ISSUE THREAD

**Issue by GrishaevPavel** [DefinitelyTyped/DefinitelyTyped#27294]
Hi. I tried this types recently, 'cause I'm new in TypeScript. And I see wrong types everywhere.
For example .attr returns not string, as we have in the d.ts, but string | undefined | number, as written in the jQuery pages. And I saw many examples, but not note them to show.

But if anybody change .attr return type, I feel that it may brake someones optimistic code. Especially if it is known, that in particular place we have just a string (and there is a question, if my knowledge is right, perhaps the error is correct).

Another example - method .offset( coordinates ). It may have both left and top, or just left or just top.

If I found this mistakes in first day I tried d.ts, it means something, but not "The repository for high quality TypeScript type definitions."

**Comment by GrishaevPavel** [comment]
Again. Look:

/**
     * Retrieve all the **elements** contained in the jQuery set, as an array.
     * @name toArray
     * @see {@link https://api.jquery.com/toArray/}
     */
    toArray(): **HTMLElement**[];

Someone do not distinguish Element and HTMLElement. This methos works with any elements, not just html, but with SVG too.

**Comment by GrishaevPavel** [comment]
> [index: number]: HTMLElement;

No! Element

**Comment by GrishaevPavel** [comment]
FaseIn must have this variant too:
fadeIn(complete?: Function): JQuery;

Because it is correct, that you have just callback.

**Comment by leonard-thieu** [comment]
It looks like you're using the v2 typings. The v3 typings have many of these issues fixed. Consider using the v3 typings if they're an option for you.

Also, as mentioned in the issue template, you should really mention the declaration authors. Due to the nature of the DefinitelyTyped repository, it's very difficult for authors to be made aware of relevant issues otherwise.

**Comment by GrishaevPavel** [comment]
Hi. Sorry, but I didn't know who I must mention, because jQuery types github page is broken, I can't see it. And in commits I see you and other developers.
Yes, I used v2, It's my fault. But in v3 I see that problems too (I am possibly wrong).
For example JQuery.Coordinates in method "offset" as an argument is a mistake. I think there must be JQuery.Coordinates-like object with optional keys.
And as I wrote about SVGElement, it seems at first sight, that TElement is HTMLElement (I am not sure). That is wrong for svg. May be HTMLElement is better for most cases. But when I write an svg app, I had to do something like "$()[0] as any as SVGElement" or change d.ts.

**Comment by leonard-thieu** [comment]
> For example JQuery.Coordinates in method "offset" as an argument is a mistake. I think there must be JQuery.Coordinates-like object with optional keys.

Will be fixed by #27437. Thank for you reporting this.

> And as I wrote about SVGElement, it seems at first sight, that TElement is HTMLElement (I am not sure). That is wrong for svg. May be HTMLElement is better for most cases. But when I write an svg app, I had to do something like "$()[0] as any as SVGElement" or change d.ts.

`TElement` can actually be set to any type. `HTMLElement` is merely the default. In some cases, TypeScript will be able to infer the element type. Otherwise, you may explicitly specify the type when creating the jQuery object.

```typescript
// $ExpectType JQuery<SVGSVGElement>
$($(document.createElementNS("http://www.w3.org/2000/svg", "svg")));

// $ExpectType JQuery<SVGLineElement>
$<SVGLineElement>('.mysvgline');
```

**Comment by GrishaevPavel** [comment]
Thank you much!
